homestar-runtime/src/event_handler/swarm_event.rs

```diff
@@ -376,6 +376,26 @@ async fn handle_swarm_event<THandlerErr: fmt::Debug + Send, DB: Database>(
             );
         }
         SwarmEvent::IncomingConnection { .. } => {}
+        SwarmEvent::ConnectionEstablished {
+            peer_id, endpoint, ..
+        } => {
+            let behavior = event_handler.swarm.behaviour_mut();
+            // only listener addresses should be added to the routing table.
+            // TODO: add identify here to discover more listen addresses and add those.
+            if endpoint.is_listener() {
+                // ignores if the peer failed or not to be added in the routing table.
+                behavior
+                    .kademlia
+                    .add_address(&peer_id, endpoint.get_remote_address().clone());
+            }
+
+            // add peer to connected peers list
+            event_handler.connected_peers.insert(peer_id, endpoint);
+        }
+        SwarmEvent::ConnectionClosed { peer_id, cause, .. } => {
+            info!("peer connection closed {peer_id}, cause: {cause:?}");
+            event_handler.connected_peers.remove_entry(&peer_id);
```

Comment:
we should also remove an entry on the OutgoingConnectionError.

Reply:
We only insert connections on successful connection anyways, so we shouldn't need to remove from the set on OutgoingConnectionError. Connections that are established and then err will route through ConnectionClosed.

Reply:
This is me basing a bit of this on https://github.com/libp2p/rust-libp2p/blob/master/examples/file-sharing/src/network.rs#L319 and reading around. There's a feeling that maybe ConnectionClosed may not always fire or fire along with the error.

Reply:
`OutgoingConnectionError` should only be triggered before `ConnectionEstablished` on an outgoing dial. Once we _have_ made a connection (and added them to connected peers) `ConnectionClosed` should be the only error type to trigger. Since we only add to connected peers on success, there shouldn't ever be a case where we need to remove a peer from `connected` in `OutgoingConnectionError`

Reply:
@mriise yep, agreed, as per docs:

> A [ConnectionEstablished](https://docs.rs/libp2p/latest/libp2p/swarm/enum.SwarmEvent.html#variant.ConnectionEstablished) event is reported if the dialing attempt succeeds, otherwise a [OutgoingConnectionError](https://docs.rs/libp2p/latest/libp2p/swarm/enum.SwarmEvent.html#variant.OutgoingConnectionError) event is reported.

Though, interesting how the example doesn't necessarily adhere to this.